app/src/main/java/com/dimagi/biometric/fragments/BaseMatchFragment.java

```diff
@@ -0,0 +1,162 @@
+        RATIONALE,
+        SETTINGS
+    }
+
+    protected abstract void handleStartCapture();
+    protected abstract void handleCancelCapture();
+
+    public BaseMatchFragment() {
+        // Required empty public constructor
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_main_menu, container, false);
+        Button startButton = view.findViewById(R.id.start_capture_button);
+        startButton.setOnClickListener(v -> {
+            if (ContextCompat.checkSelfPermission(
+                    requireContext(), Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
+                handleStartCapture();
+            } else {
+                if (shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) {
+                    savePreviousRationale();
+                }
+                if (!isPermissionDeniedPermanently()) {
+                    showPermissionAlertDialog();
+                }
+            }
+        });
+        Button cancelButton = view.findViewById(R.id.cancel_capture_button);
+        cancelButton.setOnClickListener(v -> handleCancelCapture());
+        if (savedInstanceState != null) {
+            restoreDialog(savedInstanceState);
+        }
+        return view;
+    }
+
+    private final ActivityResultLauncher<String> requestPermissionLauncher =
+            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
+        if (isGranted) {
+            handleStartCapture();
+        } else if (isPermissionDeniedPermanently()) {
+            requestOpenAppSettings();
+        }
+    });
+
+    private void showPermissionAlertDialog() {
+        dialogState = DialogState.RATIONALE;
+        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(requireContext());
+        alertDialogBuilder.setTitle(getText(R.string.camera_permission_required));
+        alertDialogBuilder.setMessage(getText(R.string.camera_permission_rationale));
+        alertDialogBuilder.setPositiveButton(getText(R.string.confirm), (dialog, which) -> requestPermissionLauncher.launch(Manifest.permission.CAMERA));
+        alertDialogBuilder.setNegativeButton(getText(R.string.cancel), null);
+        AlertDialog alertDialog = alertDialogBuilder.create();
+        alertDialog.setOnDismissListener((dialog) -> dialogState = DialogState.NONE);
+        alertDialog.show();
+    }
+
+    private void requestOpenAppSettings() {
+        dialogState = DialogState.SETTINGS;
+        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(requireContext());
+        alertDialogBuilder.setTitle(getText(R.string.go_to_settings));
+        alertDialogBuilder.setMessage(getText(R.string.camera_permission_settings));
+        alertDialogBuilder.setPositiveButton(getText(R.string.confirm), (dialogInterface, i) -> {
+            Intent intent = new Intent();
```

Comment:
Can we refactor this to a separate method?

Reply:
Refactor has been done ([8f1e995](https://github.com/dimagi/biometric-android/pull/1/commits/8f1e995908a0e55023fd53a2ba8960ad2f2acf52)).